### ISSUE TITLE
Is DeferredEvaluation actually required?

### ISSUE THREAD

**Issue by markfinal** [markfinal/BuildAMation#465]
The BAM process works like this:
- build graph
- evaluate 'out of dateness' (sync or async)
- execute the things out of date (the build)

When something in the graph is generated (e.g. source generator) by executing it, then there is a chicken-n-egg problem. How do you know to recompile a source file (in the evaluation stage) if it hasn't been written to disk yet (the build stage)?

This was previously resolved by marking certain modules as requiring 'deferred evaluation', which meant adding custom evaluation code at the build steps.

I am wondering now whether there is a better way of doing it, without requiring any additional code or logic.

In single threaded mode:
- make evaluation and build in lock-step. It's going to be doing the same amount of work on the single thread anyway, so it shouldn't be any faster, or slower. And if the dependency graph is correct, then modules will be evaluated in the correct order to determine any source generation.

In multi-threaded mode:
- all evaluation should be done in async, and any module that has procedurally generated elements, must wait on the build tasks of each of those dependents. e.g. if you come to evaluate a source file, wait on the source generator to have run, and then evaluate the source file.
